File: javascript/packages/highlighter/test/integration.test.ts
import { describe, test, expect, beforeAll, afterAll } from "vitest"
import { Herb } from "@herb-tools/node-wasm"
import { execSync } from "child_process"
import { writeFileSync, unlinkSync, mkdtempSync, rmSync } from "fs"
import { tmpdir } from "os"
import { join } from "path"

describe("herb-highlight CLI", () => {
  const testFile = join(__dirname, "test-template.html.erb")
  const normalize = (output: string) => output.replaceAll(testFile, "<test-file>")

  beforeAll(async () => {
    await Herb.load()

    writeFileSync(
      testFile,
      `<h1 class="title">
  <% if user.present? %>
    Welcome <%= user.name %>!
  <% else %>
    Please sign in
  <% end %>
</h1>`,
    )
  })

  afterAll(() => {
    try {
      unlinkSync(testFile)
    } catch {}
  })

  test("should highlight file via CLI", () => {
    const result = execSync(`node ./bin/herb-highlight ${testFile}`, {
      encoding: "utf8",
      cwd: process.cwd(),
    })

    expect(normalize(result)).toMatchSnapshot()
  })

  test("should handle non-existent file gracefully", () => {
    expect(() => {
      execSync("node ./bin/herb-highlight non-existent-file.erb", {
        encoding: "utf8",
        cwd: process.cwd(),
      })
    }).toThrow()
  })

  test("should respect NO_COLOR environment variable", () => {
    const result = execSync(
      `NO_COLOR=1 node ./bin/herb-highlight ${testFile}`,
      {
        encoding: "utf8",
        cwd: process.cwd(),
      },
    )

    expect(result).not.toContain("\x1b[")
    expect(normalize(result)).toMatchSnapshot()
  })

  test("should support --focus option", () => {
    const result = execSync(`node ./bin/herb-highlight ${testFile} --focus 3`, {
      encoding: "utf8",
      cwd: process.cwd(),
    })

    expect(normalize(result)).toMatchSnapshot()
  })

  test("should support --context-lines option", () => {
    const result = execSync(
      `node ./bin/herb-highlight ${testFile} --focus 3 --context-lines 1`,
      {
        encoding: "utf8",
        cwd: process.cwd(),
      },
    )

    expect(normalize(result)).toMatchSnapshot()
  })

  test("should handle invalid --focus value", () => {
    expect(() => {
      execSync("node ./bin/herb-highlight test-template.html.erb --focus abc", {
        encoding: "utf8",
        cwd: process.cwd(),
      })
    }).toThrow(/Invalid focus line/)
  })

  test("should require --diagnostics for --split-diagnostics", () => {
    expect(() => {
      execSync(`node ./bin/herb-highlight ${testFile} --split-diagnostics`, {
        encoding: "utf8",
        cwd: process.cwd(),
      })
    }).toThrow(/--split-diagnostics requires --diagnostics/)
  })

  test("should render diagnostics passed as JSON", () => {
    const diagnostics = JSON.stringify([{
      message: "Test offense",
      severity: "error",
      code: "test-rule",
      location: { start: { line: 1, column: 4 }, end: { line: 1, column: 9 } },
    }])

    const result = execSync(
      `NO_COLOR=1 node ./bin/herb-highlight ${testFile} --diagnostics '${diagnostics}'`,
      { encoding: "utf8", cwd: process.cwd() },
    )

    expect(normalize(result)).toMatchSnapshot()
  })

  test("should reject diagnostics that are missing required fields", () => {
    expect(() => {
      execSync(
        `node ./bin/herb-highlight ${testFile} --diagnostics '{"message":"no location"}'`,
        { encoding: "utf8", cwd: process.cwd() },
      )
    }).toThrow(/each diagnostic must have message, location, and severity/)
  })

  test("should show the usage for --help", () => {
    const result = execSync("node ./bin/herb-highlight --help", {
      encoding: "utf8",
      cwd: process.cwd(),
    })

    expect(result).toMatchSnapshot()
  })

  test("should show the versions for --version", () => {
    const result = execSync("node ./bin/herb-highlight --version", {
      encoding: "utf8",
      cwd: process.cwd(),
    })

    expect(result.startsWith("Versions:\n")).toBe(true)
    expect(result).toContain("@herb-tools/highlighter@")
  })

  test("should require an input file", () => {
    expect(() => {
      execSync("node ./bin/herb-highlight", {
        encoding: "utf8",
        cwd: process.cwd(),
      })
    }).toThrow(/Please specify an input file/)
  })
})

describe("herb-highlight CLI diff", () => {
  let directory: string

  beforeAll(async () => {
    await Herb.load()

    directory = mkdtempSync(join(tmpdir(), "herb-highlight-"))
  })

  afterAll(() => {
    try {
      rmSync(directory, { recursive: true, force: true })
    } catch {}
  })

  test("should diff two files via the diff subcommand", () => {
    const before = join(directory, "before.html.erb")
    const after = join(directory, "after.html.erb")

    writeFileSync(before, `<div>\n  <span class='card'>x</span>\n</div>`)
    writeFileSync(after, `<div>\n  <span class="card">x</span>\n</div>`)

    const result = execSync(`NO_COLOR=1 node ./bin/herb-highlight diff ${before} ${after}`, {
      encoding: "utf8",
      cwd: process.cwd(),
    })

    expect(result.replaceAll(before, "<before>").replaceAll(after, "<after>")).toMatchSnapshot()
  })

  test("should render a diff given as JSON", () => {
    const json = JSON.stringify({ original: `<img src="a.png">`, modified: `<img src="a.png" alt="">` })

    const result = execSync(`NO_COLOR=1 node ./bin/herb-highlight diff '${json}'`, {
      encoding: "utf8",
      cwd: process.cwd(),
    })

    expect(result).toMatchSnapshot()
  })

  test("should render a unified diff given as a file", () => {
    const patch = join(directory, "fix.patch")

    writeFileSync(patch, [
      "--- a/app/views/gems/index.html.erb",
      "+++ b/app/views/gems/index.html.erb",
      "@@ -1,3 +1,3 @@",
      ` <div id="gems">`,
      `-  <span class='card'>x</span>`,
      `+  <span class="card">x</span>`,
      " </div>",
      "",
    ].join("\n"))

    const result = execSync(`NO_COLOR=1 node ./bin/herb-highlight diff ${patch}`, {
      encoding: "utf8",
      cwd: process.cwd(),
    })

    expect(result).toMatchSnapshot()
  })
})

File: rust/herb-highlighter/tests/integration_test.rs
#![cfg(feature = "cli")]

use std::fs;
use std::path::Path;
use std::process::Command;

use herb_highlighter::ansi::find_ansi_sequences;

const TEMPLATE: &str = "<h1 class=\"title\">\n  <% if user.present? %>\n    Welcome <%= user.name %>!\n  <% else %>\n    Please sign in\n  <% end %>\n</h1>";

struct Fixture {
  directory: tempfile::TempDir,
}

impl Fixture {
  fn new() -> Self {
    let directory = tempfile::tempdir().expect("a temporary directory");

    fs::write(directory.path().join("test-template.html.erb"), TEMPLATE).expect("the template is written");

    Self { directory }
  }

  fn path(&self) -> String {
    self.directory.path().join("test-template.html.erb").to_string_lossy().into_owned()
  }

  fn normalize(&self, output: &str) -> String {
    output.replace(&self.path(), "<test-file>")
  }
}

fn run(arguments: &[&str], color: bool) -> std::process::Output {
  let mut command = Command::new(env!("CARGO_BIN_EXE_herb-highlight"));

  command.args(arguments);
  command.current_dir(Path::new(env!("CARGO_MANIFEST_DIR")));

  if color {
    command.env_remove("NO_COLOR");
  } else {
    command.env("NO_COLOR", "1");
  }

  command.output().expect("the CLI runs")
}

#[test]
fn highlights_a_file_via_the_cli() {
  let fixture = Fixture::new();
  let output = run(&[&fixture.path()], true);

  assert!(output.status.success());

  insta::assert_snapshot!(fixture.normalize(&String::from_utf8_lossy(&output.stdout)));
}

#[test]
fn handles_a_non_existent_file_gracefully() {
  let output = run(&["non-existent-file.erb"], true);

  assert!(!output.status.success());
  assert!(String::from_utf8_lossy(&output.stderr).contains("File not found: non-existent-file.erb"));
}

#[test]
fn respects_the_no_color_environment_variable() {
  let fixture = Fixture::new();
  let output = run(&[&fixture.path()], false);

  let stdout = String::from_utf8_lossy(&output.stdout);

  assert!(find_ansi_sequences(&stdout).is_empty());

  insta::assert_snapshot!(fixture.normalize(&stdout));
}

#[test]
fn supports_the_focus_option() {
  let fixture = Fixture::new();
  let output = run(&[&fixture.path(), "--focus", "3"], true);

  assert!(output.status.success());

  insta::assert_snapshot!(fixture.normalize(&String::from_utf8_lossy(&output.stdout)));
}

#[test]
fn supports_the_context_lines_option() {
  let fixture = Fixture::new();
  let output = run(&[&fixture.path(), "--focus", "3", "--context-lines", "1"], true);

  assert!(output.status.success());

  insta::assert_snapshot!(fixture.normalize(&String::from_utf8_lossy(&output.stdout)));
}

#[test]
fn handles_an_invalid_focus_value() {
  let fixture = Fixture::new();
  let output = run(&[&fixture.path(), "--focus", "abc"], true);

  assert!(!output.status.success());
  assert!(String::from_utf8_lossy(&output.stderr).contains("Invalid focus line"));
}

#[test]
fn requires_diagnostics_for_split_diagnostics() {
  let fixture = Fixture::new();
  let output = run(&[&fixture.path(), "--split-diagnostics"], true);

  assert!(!output.status.success());
  assert!(String::from_utf8_lossy(&output.stderr).contains("--split-diagnostics requires --diagnostics"));
}

#[test]
fn renders_diagnostics_passed_as_json() {
  let fixture = Fixture::new();

  let diagnostics =
    r#"[{"message":"Test offense","severity":"error","code":"test-rule","location":{"start":{"line":1,"column":4},"end":{"line":1,"column":9}}}]"#;

  let output = run(&[&fixture.path(), "--diagnostics", diagnostics], false);

  assert!(output.status.success());

  insta::assert_snapshot!(fixture.normalize(&String::from_utf8_lossy(&output.stdout)));
}

#[test]
fn rejects_diagnostics_that_are_missing_required_fields() {
  let fixture = Fixture::new();
  let output = run(&[&fixture.path(), "--diagnostics", r#"{"message":"no location"}"#], true);

  assert!(!output.status.success());
  assert!(String::from_utf8_lossy(&output.stderr).contains("each diagnostic must have message, location, and severity"));
}

#[test]
fn diffs_two_files_via_the_diff_subcommand() {
  let directory = tempfile::tempdir().expect("a temporary directory");

  let before = directory.path().join("before.html.erb");
  let after = directory.path().join("after.html.erb");

  fs::write(&before, "<div>\n  <span class='card'>x</span>\n</div>").expect("the file is written");
  fs::write(&after, "<div>\n  <span class=\"card\">x</span>\n</div>").expect("the file is written");

  let output = run(&["diff", &before.to_string_lossy(), &after.to_string_lossy()], false);

  assert!(output.status.success());

  let stdout = String::from_utf8_lossy(&output.stdout)
    .replace(&*before.to_string_lossy(), "<before>")
    .replace(&*after.to_string_lossy(), "<after>");

  insta::assert_snapshot!(stdout);
}

#[test]
fn renders_a_diff_given_as_json() {
  let output = run(
    &["diff", r#"{"original": "<img src=\"a.png\">", "modified": "<img src=\"a.png\" alt=\"\">"}"#],
    false,
  );

  assert!(output.status.success());

  insta::assert_snapshot!(String::from_utf8_lossy(&output.stdout));
}

#[test]
fn renders_a_unified_diff_given_as_a_file() {
  let directory = tempfile::tempdir().expect("a temporary directory");
  let patch = directory.path().join("fix.patch");

  fs::write(
    &patch,
    "--- a/app/views/gems/index.html.erb\n+++ b/app/views/gems/index.html.erb\n@@ -1,3 +1,3 @@\n <div id=\"gems\">\n-  <span class='card'>x</span>\n+  <span class=\"card\">x</span>\n </div>\n",
  )
  .expect("the patch is written");

  let output = run(&["diff", &patch.to_string_lossy()], false);

  assert!(output.status.success());

  insta::assert_snapshot!(String::from_utf8_lossy(&output.stdout));
}

#[test]
fn shows_the_usage_for_help() {
  let output = run(&["--help"], false);

  assert!(output.status.success());

  insta::assert_snapshot!(String::from_utf8_lossy(&output.stdout));
}

#[test]
fn shows_the_versions() {
  let output = run(&["--version"], false);

  assert!(output.status.success());

  let stdout = String::from_utf8_lossy(&output.stdout);

  assert!(stdout.starts_with("Versions:\n"));
  assert!(stdout.contains(&format!("herb-highlighter@{}", env!("CARGO_PKG_VERSION"))));
}

#[test]
fn requires_an_input_file() {
  let output = run(&[], false);

  assert!(!output.status.success());
  assert!(String::from_utf8_lossy(&output.stderr).contains("Please specify an input file."));
}
